Operator updates to support generic math (#29794)
dotnet/docs: docs/csharp/language-reference/operators/bitwise-and-shift-operators.md

The following operators perform bitwise or shift operations with operands of the [integral numeric types](../builtin-types/integral-numeric-types.md) or the [char](../builtin-types/char.md) type:

- Unary [`~` (bitwise complement)](#bitwise-complement-operator-) operator
- Binary [`<<` (left shift)](#left-shift-operator-), [`>>` (right shift)](#right-shift-operator-) shift, and [`>>>` (unsigned right shift)](#unsigned-right-shift-operator-) operators
- Binary [`&` (logical AND)](#logical-and-operator-), [`|` (logical OR)](#logical-or-operator-), and [`^` (logical exclusive OR)](#logical-exclusive-or-operator-) operators

> Those operators are defined for the `int`, `uint`, `long`, and `ulong` types. When both operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral types, their values are converted to the closest containing integral type. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md). The compound operators (such as `>>=`) do not convert their arguments to `int` or have the result type as `int`.

Reviewer: ```suggestion
Those operators are defined for the `int`, `uint`, `long`, and `ulong` types. When both operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral types, their values are converted to the closest containing integral type. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md). The compound operators (such as `>>=`) don't convert their arguments to `int` or have the result type as `int`.
```

The following operators perform bitwise or shift operations with operands of the [integral numeric types](../builtin-types/integral-numeric-types.md) or the [char](../builtin-types/char.md) type:

- Unary [`~` (bitwise complement)](#bitwise-complement-operator-) operator
- Binary [`<<` (left shift)](#left-shift-operator-), [`>>` (right shift)](#right-shift-operator-) shift, and [`>>>` (unsigned right shift)](#unsigned-right-shift-operator-) operators
- Binary [`&` (logical AND)](#logical-and-operator-), [`|` (logical OR)](#logical-or-operator-), and [`^` (logical exclusive OR)](#logical-exclusive-or-operator-) operators

> Those operators are defined for the `int`, `uint`, `long`, and `ulong` types. When both operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral types, their values are converted to the closest containing integral type. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md).

Reviewer 2: Unrelated but this statement is a bit confusing I think. 

`byte >> int` is treated as `int = ((int)byte) >> int` but `byte >>= int` works and is treated as `byte = byte >> int`. Same goes for the other small types.

Author: I made a similar change as the text in the arithmetic operators article.

The high-order empty bit positions are set based on the type of the left-hand operand as follows:

- If the left-hand operand is of type `int` or `long`, the right-shift operator performs an *arithmetic* shift: the value of the most significant bit (the sign bit) of the left-hand operand is propagated to the high-order empty bit positions. That is, the high-order empty bit positions are set to zero if the left-hand operand is non-negative and set to one if it's negative.

  :::code language="csharp-interactive" source="snippets/shared/BitwiseAndShiftOperators.cs" id="ArithmeticRightShift":::

- If the left-hand operand is of type `uint` or `ulong`, the right-shift operator performs a *logical* shift: the high-order empty bit positions are always set to zero.

  :::code language="csharp-interactive" source="snippets/shared/BitwiseAndShiftOperators.cs" id="LogicalRightShift":::

> [!NOTE]
> Use the [unsigned right-shift operator](#unsigned-right-shift-operator-) to perform a *logical* shift on signed integer types. This is preferred to casting the signed type to an unsigned type and casting back after the shift.

## Unsigned right-shift operator >>>

> The `>>>` operator, available starting in C# 11, performs a *logical* shift on all integer types. The high-order bits are always set to zero, regardless of the type of the left side argument. The following example demonstrates the difference between `>>` and `>>>` for a negative integer value:

Reviewer 2: This looks to be the first introduction of `*logical*` shift. There seems to be no mention of `*arithmetic*` anywhere either or that unsigned types always use a `*logical*` shift.
